Event received_at / occurred_at definition (#659)
zalando/restful-api-guidelines: chapters/events.adoc

      type: string
      example: 'example.important-business-event'
    occurred_at:
      description: |
         Timestamp of when the event was processed and created by the producer application. 
         (Remark: `created_at` or `processed_at` would be even more precise namings.)

>          The timestamp may differ from when the actual business event happened, e.g. 'packet delivered to the customer'. 

Reviewer: ```suggestion
         The timestamp may differ from when the actual business event happened, e.g. `packet delivered to the customer`. 
```

Author: 👍

>          In these situations we expect that the information is passed via an additional event type specific attribute e.g. 'delivered_at'. 

Reviewer: ```suggestion
         In these situations we expect that the information is passed via an additional attribute in the event-type, e.g. 'delivered_at'. 
```

Author: 👍

>          (Hint: `occurred_at` timestamp also typically differs from when the event is posted by the producer and received by Nakadi via the API.) 

Reviewer: i dont understand this hint. can someone clarify/reword/explain at least here in a comment what this means?

Author: Updated, to make it (hopefully) clearer.

      type: string
      format: date-time
      example: '1996-12-19T16:39:57-08:00'
    received_at:
      description: |
        Timestamp of when the event was received by Nakadi via the event post endpoints. 

>         It is automatically enriched, and events will be rejected if set by the Event producer.

Reviewer: ```suggestion
        It is automatically enriched, and events will be rejected if set by the event producer.
```

Author: 👍

      type: string
      example: 'example.important-business-event'
    occurred_at:
      description: |
         Timestamp of when the event was processed and created by the producer application. 
         (Remark: `created_at` or `processed_at` would be even more precise namings.)
         This timestamp may differ from when the actual business event happened, for instance, when the packet was delivered to the customer. 
         In these situations we expect that the information is passed via an additional attribute in the event type e.g. `delivered_at`. 
         Hint: Depending on the producer implementation, this `occurred_at` timestamp typically is some milliseconds earlier than 

>          the event publication and receival by Nakadi servers via the API post endpoints -- see below.

Author: ok, I will introduce further line breaks. 

Wrt. other notes: This is a lengthy discussion, not necessarily helpful for the concrete attribute description. 
Target was to address the continuous uncertainties around semantics of 'occurred_at' -- see e.g. https://chat.google.com/room/AAAALJjUd9k/zDnRj47g8Jg -- and to make clear that there may by a difference between actual business event vs. time of event object created and processed. 
Minor update done.

>          the event publication and receival by Nakadi servers via the API post endpoints -- see below.

Author: I will provide update wrt. to the event processing aspect, as discussed

      type: string
      format: date-time
      example: '1996-12-19T16:39:57-08:00'
    received_at:
      description: |

>         Timestamp of when the event was received by Nakadi via the event post endpoints. 

Author: 1) It IMO time to reconsider our ambition to define event standards outside Nakadi. However, I don't want to discuss it in context of this issue and slightly modified it. 
2) Proposal definition is more precise. (In fact, broker example above is wrong in Nakadi case)

>         It is automatically enriched, and events will be rejected if set by the event producer.

Author: Exactly, rejection is worth to be mentioned in the description.

>         It is automatically enriched, and events will be rejected if set by the event producer.

Reviewer 2: This is true only for properties in `metadata` object. This schema is called _The core Event Type_ which doesn't match to any schema supported by Nakadi.

      type: string
      example: 'example.important-business-event'
    occurred_at:
      description: |
         Technical timestamp of when the event object was created during processing of the
         business event by the producer application. Note, it may differ from the timestamp 
         when the related real-world business event happened (e.g. when the packet was handed 
         over to the customer), which should be passed separately via an event type specific 
         attribute. 

>          Depending on the producer implementation, the timestamp is typically some milliseconds  

Reviewer 2: Which `occurred_at` field is this? If it is the [field in the `metadata`](https://nakadi.io/manual.html#definition_EventMetadata*occurred_at), then it has a `date-time` format.

Author: Pls. help me to understand your comment.
Yes, this `occurred_at` field is an event metadata attribute defined with format `date-time`. 
Do you mean using `timestamp` wording is misleading, since the date has `date-time` format?
Timestamps can be provided in different formats and `date-time` (based on RFC 3339) is our requested standard format.

Reviewer 2: @tfrauenstein sorry, I has overlooked that this is already in `metadata` field. All good from my side.

      type: string
      example: 'example.important-business-event'
    occurred_at:
      description: |
         Timestamp of when the event object was created and processed by the producer application. 
         (Remark: `created_at` or `processed_at` would be even more precise namings.)
         This timestamp may differ from when the actual business event happened, for instance, 
         when the packet was delivered to the customer. 
         In these situations we expect that the information is passed via an additional 

>          attribute in the event type e.g. `delivered_at`. 

Reviewer 3: I still have my concerns about mentioning `processed_at` and `delivered_at` here, because these are not related to the creation of the event, that is supposed to be addressed by `occurred_at`, but to the processing of a possibly existing parent event. Since I cannot see how this contributes to make the semantic of `occurred_at` clearer, I suggest to remove the marked to sentences.

Author: thx. I will merge your contribution (w/o received_at) incl. processing aspect, as discussed
